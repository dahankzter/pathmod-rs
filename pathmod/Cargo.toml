[package]
name = "pathmod"
version = "0.1.0"
edition = "2021"
license = "MIT OR Apache-2.0"
description = "Re-export crate for pathmod"
repository = ""

[dependencies]
pathmod_core = { path = "../pathmod_core" }
pathmod_derive = { path = "../pathmod_derive" }

[lib]
path = "src/lib.rs"
