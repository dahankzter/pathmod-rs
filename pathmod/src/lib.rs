//! Pathmod â€” tiny, composable accessors for ergonomic, type-safe deep mutation.
//!
//! This is the user-facing re-export crate. Depend on `pathmod` and bring the
//! prelude into scope to use both the derive macros and the runtime:
//!
//! ```rust
//! use pathmod::prelude::*;
//!
//! #[derive(Accessor, Debug, PartialEq)]
//! struct Bar { x: i32 }
//!
//! #[derive(Accessor, Debug, PartialEq)]
//! struct Foo { a: i32, b: Bar }
//!
//! let mut foo = Foo { a: 1, b: Bar { x: 2 } };
//! // Access direct field
//! let acc_a = Foo::acc_a();
//! assert_eq!(*acc_a.get(&foo), 1);
//! acc_a.set(&mut foo, 10);
//! assert_eq!(foo.a, 10);
//!
//! // Compose to reach nested leaf Foo -> Bar -> x
//! let acc_bx = Foo::acc_b().compose(Bar::acc_x());
//! acc_bx.set_mut(&mut foo, |v| *v += 5);
//! assert_eq!(foo.b.x, 7);
//! ```
//!
//! See also:
//! - [`pathmod_core`] for the `Accessor` runtime and its operations.
//! - [`pathmod_derive`] for the derive macros generating `acc_*` and `with_*`.

pub use pathmod_core::*;
pub use pathmod_derive::*;

/// Prelude re-exporting the most commonly used items.
///
/// Bring this into scope in your modules:
///
/// ```rust
/// use pathmod::prelude::*;
/// ```
pub mod prelude {
    pub use pathmod_core::prelude::*;
    pub use pathmod_derive::*;
}
