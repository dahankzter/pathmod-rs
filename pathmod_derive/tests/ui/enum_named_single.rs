use pathmod::prelude::*;

#[derive(EnumAccess)]
enum V { Named { v: i32 } }

fn main() {}
