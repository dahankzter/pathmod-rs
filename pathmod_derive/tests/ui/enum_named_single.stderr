error: #[derive(EnumAccess)] currently supports only tuple variants with exactly one field; named-field single variants are not yet supported
 --> tests/ui/enum_named_single.rs:3:10
  |
3 | #[derive(EnumAccess)]
  |          ^^^^^^^^^^
  |
  = note: this error originates in the derive macro `EnumAccess` (in Nightly builds, run with -Z macro-backtrace for more info)
